### ISSUE TITLE
Unexpected behaviour when using BasisStatePreparation after other gates

### ISSUE THREAD

**Issue by mariaschuld** [PennyLaneAI/pennylane#1191]
Consider the following example:

``` python

@qml.qnode(dev)
def circuit():
    qml.PauliX(wires=(0))
    qml.templates.state_preparations.basis.BasisStatePreparation([1,0,0], wires=(0,1,2))
    return qml.state()
```

* *Expected behavior:* (What you expect to happen)

The user would expect that after a state preparation routine, the state is in `|1,0, 0>`. Note that this is in some sense unphysical (since it disposes of the state before state preparation), and a better behaviour would be to throw an error.

* *Actual behavior:* (What actually happens)

At the end of the circuit, the state is in `|000>`, since the `BasicStatePreparation` operation just applies `PauliX` gates - which turns the first qubit back to 0.

* *Proposed short-term fix:*

Register `BasicStatePreparation` with the [list of operations](https://github.com/PennyLaneAI/pennylane/blob/fd99be754d55bbb919aadbbbdff70e40fbe3bcbf/pennylane/tape/tape.py#L31) that have to be applied first in a circuit.

Note that this is a breaking change!

* *Long-term discussion:*

We should revisit our system of considering operations as "first-only". For example, why are other state preparation templates not forced to be first in a circuit? Shall we impose this restriction at all?

**Comment by albi3ro** [comment]
On further thought, what if a user wanted to apply one state preparation on a portion of the the wires, and a different state preparation on a different section of the wires?
They wouldn't overlap and would both be state preparations, just one would occur after the other. A state preparation only needs to be the first operator on _those wires_ , not on the device as a whole.

**Comment by mariaschuld** [comment]
Yes, true. This discussion is actually a larger issue we had before - to make the "first operation" check consider the wires. But that would not only touch this template.

**Comment by co9olguy** [comment]
One thing to think about here (that we visited before in SF) is that it is relatively straightforward to do the bookkeeping to determine whether some wires have already been "used" before (and hence valid targets for state preparations), but doing so may introduce some extra overhead (whose magnitude depends on the implementation details).

Generally I'm in favour of allowing state preps to be more flexible (providing the bookkeeping is not introducing overead), not just purely as a user convenience, but also because we now have qubit resets on HW and plan future support in PL for measurement feedforward

**Comment by astralcai** [comment]
nvm this is still wrong

**Comment by dwierichs** [comment]
@astralcai I would second that this is actually fine now. By now we do not have the original strict requirements anymore about state preparation operations, and the convention basically is to expect a unitary operation to happen that 
- would prepare the requested state when applied to the state $|0\rangle$ and
- does not come with any guarantees about the result when applied to another state.

At least that is my understanding of current state prep in PL  🤔 

For example, `MottonenStatePreparation` does shenanigans similar to the above:
```python 
import pennylane as qml

dev = qml.device('default.qubit', wires=2)

@qml.qnode(dev)
def circuit():
    qml.PauliX(0)
    qml.MottonenStatePreparation([0.5, 0.5, 0.5, 0.5], wires=(0, 1))
    return qml.state()
circuit()
>>> array([-0.5+0.j, -0.5+0.j,  0.5+0.j,  0.5+0.j]) # These are not the signs you are looking for.
```
